### PR TITLE
[FIX] Symbol addresses with ASLR.

### PR DESCRIPTION
**Pull request by megamisan**
On modern platforms, binaries are compiled with PIE (Position Independent Executable) feature activated. Also, on recent kernels, the Address Space Layout Randomization (ASLR) is activated.
Thus, the symbol adresses passed to instrumentation functions are not the real symbol adresses in the binaries and the lookup done by `addr2line` will fail.

This change removes the offset from the adresses.

As a freebie, launching `gen_report.sh` from any location has been fixed.

As a side note, the setting up of the trace from a function address will not work.

### CODE REVIEW

**Comment by finaldie** [comment]
Hi @megamisan,

Thank you so much for looking into this ASLR issue for `ftracer`, yes it's a known issue for the tool.

As far as I know, there are some approaches can solve it:
- Disable ASLR, set 0 to `/proc/sys/kernel/randomize_va_space`, but the downside is: it requires `sudo` permission on the target host, also it will affect the environment for the following commands.
- Remove the offset while logging. This is exactly your solution in this PR, I love this idea very much.
- Read the offset information from `/proc/$pid/maps` and adjust the offset in reporting scripts. This is working well for those apps still running, otherwise, we have to dump the offset to somewhere before the app dead.
- ...

Previously, I just temporarily disable the ASLR on my testing environment, that's easier but not the best way to handle it.

And yes, as you mentioned, `function address entry` tracing doesn't work, we can think about to enhance this later, although, it right now not that perfect which only works for the main app code (doesn't work for those dynamic library function address very well). Basically, we need `<app/libname, function address in ELF, function runtime offset>` in order to restore it back to the correct function name in the reporting phase anyway.

I do have other few comments, will leave them in below as well. 

Thank you so much, very appreciated!

File: example/c/run.sh [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/bin/sh
 
-export LD_PRELOAD=../../src/ftracer.so
```
Comment (finaldie) [inline example/c/run.sh]: [Not a blocker] I see the `libdl` has already been built as a dependency, should we still append it here?

```bash
xxx@finalarch: ~/code/github/ftracer/src>ldd -r ftracer.so
	linux-vdso.so.1 (0x00007ffda7546000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f550eab6000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f550e8f0000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f550eaed000)
```

File: tools/gen_report.sh [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 #set -x
 
+curdir="$(cd "$(dirname "$0")"; pwd)"
```
Comment (finaldie) [inline tools/gen_report.sh]: :+1:

File: example/c/run.sh (hunk 1 above)
Comment (megamisan) [reply example/c/run.sh]: When I tried running it first, without adding `libdl` to the preload, it just complained the symbol `dladdr1` cannot be found.

Comment (finaldie) [reply example/c/run.sh]: I cannot reproduce your issue from my side, the issue should only occur when the tracer.so doesn't link with `ldl` while we are running the app. The purpose of the `preload` is for solving the dependency issues, link it to `tracer.so` does :)

If you have time, could you help to have a quick verification again from your environment with the latest code? Very appreciated!

**Comment by finaldie** [comment]
Thanks, @megamisan :+1:

**Pull request merged**